pkg/virtualfs/directory.go

```diff
@@ -0,0 +1,169 @@
+			name: name,
+			mode: permissions,
+			// TODO: Add owner information
+		},
+		source: func() (ReaderSeekerCloser, error) {
+			return streamReader(sourceEndpoint)
+		},
+	}
+
+	err = d.addChild(file)
+
+	return file, errors.Wrap(err, "Failed to add file")
+}
+
+// AddDir adds a fake directory with a given name and permissions.
+func (d *Directory) AddDir(name string, permissions os.FileMode) (subdir *Directory, err error) {
+	d, name, err = d.resolveSubdir(name)
+	if err != nil {
+		return nil, err
+	}
+
+	subdir = &Directory{
+		dirEntry: dirEntry{
+			name: name,
+			mode: permissions | os.ModeDir,
+		},
+	}
+
+	err = d.addChild(subdir)
+
+	return subdir, err
+}
+
+// Subdir finds a subdirectory with the given name.
+func (d *Directory) Subdir(name ...string) (*Directory, error) {
+	i := d
+
+	for _, n := range name {
+		i2 := i.children.FindByName(n)
+		if i2 == nil {
+			return nil, errors.New(fmt.Sprintf("'%s' not found in '%s'", n, i.Name()))
+		}
+
+		if !i2.IsDir() {
+			return nil, errors.New(fmt.Sprintf("'%s' is not a directory in '%s'", n, i.Name()))
+		}
+
+		i = i2.(*Directory)
+	}
+
+	return i, nil
+}
+
+// Remove removes directory dirEntry with the given name.
+func (d *Directory) Remove(name string) {
+	newChildren := d.children[:0]
+
+	for _, e := range d.children {
+		if e.Name() != name {
+			newChildren = append(newChildren, e)
+		}
+	}
+
+	d.children = newChildren
+}
+
+// OnReaddir invokes the provided function on read.
+func (d *Directory) OnReaddir(cb func()) {
+	d.onReaddir = cb
+}
+
+// Child gets the named child of a directory.
+func (d *Directory) Child(ctx context.Context, name string) (fs.Entry, error) {
+	return fs.ReadDirAndFindChild(ctx, d, name)
+}
+
+// Readdir gets the contents of a directory.
+func (d *Directory) Readdir(ctx context.Context) (fs.Entries, error) {
+	if d.readdirError != nil {
+		return nil, d.readdirError
+	}
+
+	if d.onReaddir != nil {
+		d.onReaddir()
+	}
+
+	return append(fs.Entries(nil), d.children...), nil
+}
+
+func (d *Directory) addChild(e fs.Entry) error {
+	if strings.Contains(e.Name(), "/") {
+		return errors.New("Failed to add child entry: name cannot contain '/'")
+	}
+
+	d.children = append(d.children, e)
+	d.children.Sort()
+	return nil
+}
+
+func (d *Directory) resolveSubdir(name string) (parent *Directory, leaf string, err error) {
+	parts := strings.Split(name, "/")
+	for _, n := range parts[0 : len(parts)-1] {
+		if d, err = d.Subdir(n); err != nil {
+			return nil, "", errors.Wrap(err, fmt.Sprintf("Failed to resolve sub directory '%s'", n))
+		}
+	}
+
+	return d, parts[len(parts)-1], nil
+}
+
+func (d *Directory) resolveAndAddSubdir(name string) (parent *Directory, leaf string, err error) {
+	parts := strings.Split(name, "/")
+	for _, n := range parts[0 : len(parts)-1] {
+		if d, err = d.AddDir(n, 0777); err != nil {
+			return nil, "", errors.Wrap(err, fmt.Sprintf("Failed to add sub directory '%s'", n))
+		}
+	}
+
+	return d, parts[len(parts)-1], nil
+}
```

Comment:
You could replace `resolveAndAddSubdir` with a more general `AddAllDirs` method with behavior similar to `mkdir -p`. That can be leveraged in the helper method mentioned above.

Then, you can also remove the `resolveSubdir` and `Subdir` methods.


```suggestion
// AddAllDirs creates under d all the necessary directories in pathname, similar
// to os.MkdirAll
func (d *Directory) AddAllDirs(pathname string, permissions os.FileMode) (parent *Directory, err error) {
	p, missing, err := d.resolveDirs(pathname)
	if err != nil {
		return nil, err
	}

	for _, n := range missing {
		if p, err = p.AddDir(n, permissions); err != nil {
			return nil, errors.Wrap(err, fmt.Sprintf("Failed to add sub directory '%s'", n))
		}
	}

	return p, nil
}

func (d *Directory) resolveDirs(pathname string) (parent *Directory, missing []string, err error) {
	if pathname == "" {
		return d, nil, nil
	}

	p := d
	parts := strings.Split(path.Clean(pathname), "/")

	for i, n := range parts {
		i2 := p.children.FindByName(n)
		if i2 == nil {
			return p, parts[i:], nil
		}

		if !i2.IsDir() {
			return nil, nil, errors.New(fmt.Sprintf("'%s' is not a directory in '%s'", n, p.Name()))
		}

		p = i2.(*Directory)
	}

	return p, nil, nil
}
```